kbs/src/api/src/policy_engine/opa/mod.rs

```diff
@@ -51,105 +33,211 @@ impl PolicyEngineInterface for Opa {
-        let decision_str: &CStr = unsafe { CStr::from_ptr(decision_buf) };
-        let res = decision_str.to_str()?.to_string();
-        log::debug!("Evaluated: {}", res);
-        if res.starts_with("Error::") {
-            return Err(anyhow!(res));
-        }
-
-        let res_kv: Value = serde_json::from_str(&res)?;
-        let result_boolean = res_kv["allow"]
-            .as_bool()
-            .ok_or_else(|| anyhow!("Policy Engine output must contain \"allow\" boolean value"))?;
-
-        Ok((result_boolean, res))
+    ) -> Result<bool, ResourcePolicyError> {
+        let mut engine = regorus::Engine::new();
+
+        // Add policy as data
+        engine
+            .add_policy_from_file(self.policy_path.clone())
+            .map_err(|_| ResourcePolicyError::PolicyLoadError)?;
+
+        // Add resource path as data
+        let resource_path_object =
+            regorus::Value::from_json_str(&format!("{{\"resource-path\":\"{}\"}}", resource_path))
+                .map_err(|_| ResourcePolicyError::ResourcePathError)?;
+
+        engine
+            .add_data(resource_path_object)
+            .map_err(|_| ResourcePolicyError::DataLoadError)?;
+
+        // Add TCB claims as input
+        engine
+            .set_input_json(&input_claims)
+            .map_err(|_| ResourcePolicyError::InputError)?;
+
+        engine
+            .eval_bool_query("data.policy.allow".to_string(), false)
+            .map_err(|_| ResourcePolicyError::EvaluationError)
     }
 
-    async fn set_policy(&mut self, policy: String) -> Result<()> {
+    async fn set_policy(&mut self, policy: String) -> Result<(), ResourcePolicyError> {
         let policy_bytes = base64::engine::general_purpose::URL_SAFE_NO_PAD
             .decode(policy)
-            .map_err(|e| anyhow!("Base64 decode OPA policy string failed: {:?}", e))?;
+            .map_err(|_| ResourcePolicyError::DecodeError)?;
 
         tokio::fs::write(&self.policy_path, policy_bytes)
             .await
-            .map_err(|e| anyhow!("Write OPA policy to file failed: {:?}", e))
+            .map_err(|_| ResourcePolicyError::IOError)?;
+
+        Ok(())
     }
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
+
     use base64::engine::general_purpose::URL_SAFE_NO_PAD;
+    use rstest::rstest;
     use serde_json::json;
+    use tempfile::{NamedTempFile, TempDir};
 
     fn dummy_input(product_id: &str, svn: u64) -> String {
         json!({
+            "tee": "sample",
             "tee-pubkey": "dummy-key",
             "tcb-status": {
-                "productId": product_id.to_string(),
-                "svn": svn.to_string(),
+                "productId": product_id,
+                "svn": svn
             }
         })
         .to_string()
     }
 
+    async fn set_policy_from_file(opa: &mut Opa, path: &str) -> Result<(), ResourcePolicyError> {
+        let policy = std::fs::read(PathBuf::from(path.to_string())).unwrap();
```

Comment:
I wouldn't resort to unwrap if the fn already returns a Result

Reply:
This is a helper for a test. I pass up the error from `set_policy` so that it can be checked in the tests but if we fail to read the policy file specified in the test itself I figure we should just explode immediately.